### ISSUE TITLE
Linchpin should support passing arguments directly to the underlying ansible playbooks

### ISSUE THREAD

**Issue by jaypoulz** [CentOS-PaaS-SIG/linchpin#375]
Today I spoke to @arilivigni and Jeff Burke about the virtues of linchpin as a tool. We agreed that one of the features that could greatly enhance the configurability and power of linchpin would be to the CLI an option to pass extra-args to the underlying ansible-playbooks. 

I believe this should be assessed as a potential future component of linchpin, since this would leverage the value of using ansible (configurability) while still allowing for the CLI to remain simple and easy to understand. Admittedly, this is a power user feature, but issue #155 seems like a direct example of when this kind of thing would be useful.

**Comment by herlo** [comment]
Issue #155 is a different concept altogether. Become is a specific feature in ansible. It's used to elevate permissions from a standard user to the root user, for instance. I don't think that is exactly what you are looking for in this situation. 

I do agree that passing in arguments through an --extra_vars might be a good way to go about things. Let's flesh this out a bit more and come back with a better concept of passing args to the topology and such.

**Comment by arilivigni** [comment]
@herlo I think after you finish the work for uniqueness of instances this should be the next top priority since it is something everyone can leverage and it makes linch-pin more flexible.

**Comment by herlo** [comment]
After talking with @jaypoulz @arilivigni, there are two essential parts for this feature. Each will be described in its own comment below.

**Comment by herlo** [comment]
The first is to allow dynamic PinFile, topology, and layout files. Essentially, the idea would be that a user could pass in a script in place of these items. The result of the script would be a json output, probably to STDOUT, that linchpin could ingest. 

Probably something like:

```
[{
    "topology_data": {
        "resource_groups": [{
            "credentials": {
                "filename": "clouds.yaml",
                "profile": "ci-rhos"
            },
            "resource_definitions": [{
                "count": 1,
                "fip_pool": "10.8.240.0",
                "flavor": "m1.small",
                "image": "CentOS-7-x86_64-GenericCloud-1612",
                "keypair": "ci-factory",
                "name": "database",
                "networks": [
                    "atomic-e2e-jenkins-test"
                ],
                "type": "os_server"
                },
                {
                "count": 1,
                "fip_pool": "10.8.240.0",
                "flavor": "m1.small",
                "image": "Fedora-Cloud-Base-26-compose-latest",
                "keypair": "ci-factory",
                "name": "frontend",
                "networks": [
                    "atomic-e2e-jenkins-test"
                ],
                "type": "os_server"
                }
            ],
            "resource_group_name": "os-server",
            "resource_group_type": "openstack"
        }],
        "topology_name": "os-single"
    }
},
{
    "layout_data": {
        "inventory_layout": {
            "hosts": {
                "db-node": {
                    "count": 1,
                    "host_groups": [
                        "database"
                    ]
                },
                "frontend-node": {
                    "count": 1,
                    "host_groups": [
                        "frontend"
                    ]
                }
            },
            "vars": {
                "hostname": "__IP__"
            }
        }
    }
}]
```

**Comment by herlo** [comment]
The second component would be to take the existing topologies, and allow for templating them in some way. The idea being that it would introduce a looping component and variable replacement similar to what Ansible does.

A simple use case for this might be to have three nodes in the same resource_group, but each of them be a different architecture. The architectures could either be passed in as an external variable (`{{ arches }}`) or it could be used in a loop at the bottom of the resource group. 

```
---
topology_name: os-test
resource_groups:
  - resource_group_name: lp-test
    resource_group_type: openstack
    resource_definitions:
      - name: resource
        type: os_server
        flavor: m1.small
        image: {{ image }}
        count: 3
        keypair: ci-factory
        networks:
          - atomic-e2e-jenkins-test
        fip_pool: 10.8.172.0/22
    credentials:
      filename: clouds2.yaml
      profile: ci-osp
    with_images:
      - rhel-6.4-x86_64-released
      - rhel-6.4-i686-released
      - rhel-6.4-ppc64-released
```

The validation in this case might require that the items and the count match. This will take some more thinking and might not be the end result.

**Comment by herlo** [comment]
Opening two issues for the features listed above. These will better describe and track the changes for each issue.

#402 - [Dynamic Inputs to LinchPin](https://github.com/CentOS-PaaS-SIG/linchpin/issues/375#issuecomment-342873894)
#403 - [Topology Templating](https://github.com/CentOS-PaaS-SIG/linchpin/issues/375#issuecomment-342877095l)

**Comment by herlo** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [[linchpin-dev] EPIC Dynamic Input and Templating](https://trello.com/c/nNfNmA05/253-linchpin-dev-epic-dynamic-input-and-templating)

**Comment by herlo** [comment]
Closed by #422
